### PR TITLE
PXF-6.x/ADBDEV-3096: PXF JDBC: Allow set Oracle parallel instructions

### PR DESCRIPTION
**Pull request by RomaZe**

### CODE REVIEW

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package org.greenplum.pxf.plugins.jdbc.utils.oracle;
+
+public class OracleParallelSessionParam {
+    private Clause clause;
+    private StatementType statementType;
+    private String degreeOfParallelism;
+
+    public enum Clause {
+        ENABLE("ENABLE"),
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java]: ENABLE,
        DISABLE,
        FORCE;

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        DISABLE("DISABLE"),
+        FORCE("FORCE");
+
+        public final String value;
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java]: unnecessary

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        Clause(String value) {
+            this.value = value;
+        }
+    }
+
+    public enum StatementType {
+        DML("DML"),
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java]: Same

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        DDL("DDL"),
+        QUERY("QUERY");
+
+        public final String value;
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java]: Same

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        params.put("statement_type", values[1]);
+        if (values.length == 3 && Strings.isNotBlank(values[2])) {
+            params.put("degree_of_parallelism", values[2]);
+        } else {
+            params.put("degree_of_parallelism", "");
+        }
+        return params;
+    }
+
+    private boolean isClauseValid(String value) {
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: Unnecessary

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        return Arrays.stream(OracleParallelSessionParam.Clause.values()).anyMatch(e -> e.value.equalsIgnoreCase(value));
+    }
+
+    private boolean isStatementTypeValid(String value) {
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: Unnecessary

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java (hunk 1 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java]: Fixed

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java (hunk 2 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java]: Removed

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java (hunk 3 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java]: Fixed

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java (hunk 4 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParam.java]: Removed

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java (hunk 5 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: Refactored

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java (hunk 6 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: Refactored

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        return param;
+    }
+
+    private void validateValue(String property, String value) {
+        if (StringUtils.isBlank(value)) {
+            throw new IllegalArgumentException(String.format(
+                    "The parameter '%s' is empty in jdbc-site.xml", property)
+            );
+        }
+        if (value.split(ORACLE_JDBC_SESSION_PARALLEL_PROPERTY_DELIMITER).length < 2
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: `value.split(ORACLE_JDBC_SESSION_PARALLEL_PROPERTY_DELIMITER)` - introduce method argument

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            throw new IllegalArgumentException(String.format(
+                    "The parameter '%s' in jdbc-site.xml has to contain at least 2 but not more then 3 values delimited by %s",
+                    property, ORACLE_JDBC_SESSION_PARALLEL_PROPERTY_DELIMITER)
+            );
+        }
+    }
+
+    private HashMap<String, String> getParallelSessionParam(String value) {
+        HashMap<String, String> params = new HashMap<>();
+        String[] values = value.split(ORACLE_JDBC_SESSION_PARALLEL_PROPERTY_DELIMITER);
```
Comment (iamlapa) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: `value.split(ORACLE_JDBC_SESSION_PARALLEL_PROPERTY_DELIMITER)` - introduce method argument

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java (hunk 7 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: Moved ORACLE_JDBC_SESSION_PARALLEL_PROPERTY_DELIMITER to the OracleSessionQueryFactory class. So, the ORACLE_JDBC_SESSION_PARALLEL_PROPERTY_PREFIX and ORACLE_JDBC_SESSION_PARALLEL_PROPERTY_DELIMITER in the same class now.

Introduced additional argument  - delimiter

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java (hunk 8 above)
Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: Introduced additional argument  - delimiter

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java (hunk 7 above)
Comment (iamlapa) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: i meant it:
```
    private HashMap<String, String> getParallelSessionParam(String[] values) {
        HashMap<String, String> params = new HashMap<>();
        params.put("clause", values[0]);
        params.put("statement_type", values[1]);
        if (values.length == 3 && Strings.isNotBlank(values[2])) {
            params.put("degree_of_parallelism", values[2]);
        }
        return params;
    }
```

Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: Done

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java (hunk 8 above)
Comment (iamlapa) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: I meant it:
```java
 private HashMap<String, String> getParallelSessionParam(String[] values) {
....
}
```

Comment (RomaZe) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/utils/oracle/OracleParallelSessionParamFactory.java]: Changed it and added additional function to split value with validation.

**Review by iamlapa — APPROVED** [review]

**Pull request merged**